allennlp/commands/train.py

```diff
@@ -126,31 +127,39 @@ def train_model(params: Params, serialization_dir: str) -> Model:
     logger.info("Reading training data from %s", train_data_path)
     train_data = dataset_reader.read(train_data_path)
 
-    all_datasets: List[Dataset] = [train_data]
-    datasets_in_vocab = ["train"]
+    all_datasets: Dict[str, Dataset] = {"train": train_data}
 
     validation_data_path = params.pop('validation_data_path', None)
     if validation_data_path is not None:
         logger.info("Reading validation data from %s", validation_data_path)
         validation_data = dataset_reader.read(validation_data_path)
-        all_datasets.append(validation_data)
-        datasets_in_vocab.append("validation")
+        all_datasets["validation"] = validation_data
     else:
         validation_data = None
 
     test_data_path = params.pop("test_data_path", None)
     if test_data_path is not None:
         logger.info("Reading test data from %s", test_data_path)
         test_data = dataset_reader.read(test_data_path)
-        all_datasets.append(test_data)
-        datasets_in_vocab.append("test")
+        all_datasets["test"] = test_data
     else:
         test_data = None
 
-    logger.info("Creating a vocabulary using %s data.", ", ".join(datasets_in_vocab))
+    # powerset (all permutations of all lengths) of the 3 possible options
+    # with the one containing the largest allowed set of keys given the datasets
+    # that were passed in the first position, which we use by default.
+    allowed_combinations = itertools.chain.from_iterable(itertools.permutations(all_datasets.keys(), i)
+                                                         for i in [3, 2, 1])
+
+    datasets_for_vocab_creation = params.pop_choice("datasets_for_vocab_creation",
+                                                    [list(x) for x in allowed_combinations],
+                                                    default_to_first_choice=True)
```

Comment:
on one hand this is clever, on the other hand it's almost too clever. why not just

```
datasets_for_vocab_creation = set(params.pop("datasets_for_vocab_creation", all_datasets))

for dataset in datasets_for_vocab_creation:
    if dataset not in all_datasets:
        raise ConfigurationError(f"invalid 'dataset_for_vocab_creation' {dataset}")
```

Reply:
If you haven't specified a test or evaluation dataset, it isn't a valid choice for a dataset to build the vocabulary from, though.

Reply:
doesn't my suggestion raise in that case?

Reply:
oh right, sorry. I'll change it